src/typechecker.erl

```diff
@@ -567,20 +568,37 @@ glb_ty({type, _, record, _}, {type, _, record, _}, _A, _TEnv) ->
 
 %% Map types. These are a bit tricky.
 %% For now going with a very crude approximation.
-glb_ty(Ty1 = {type, _, map, Assocs1}, Ty2 = {type, _, map, Assocs2}, _A, _TEnv) ->
+glb_ty(Ty1 = {type, _, map, Assocs1}, Ty2 = {type, _, map, Assocs2}, A, TEnv) ->
     case {Assocs1, Assocs2} of
         {any, _} -> ret(Ty2);
         {_, any} -> ret(Ty1);
         _ ->
             %% TODO: Too simplistic!
-            %% This is only accurate if Assocs1 and Assocs2 are disjoint.
-            %% Moreover, the result of `glb_ty' should be independent
-            %% of `Ty1' and `Ty2' argument order.
-            %% OTOH, just sorting the Assocs below is not possible,
-            %% since map association order is meaningful.
+            %% Map keys are used as is, not merged together even if it is possible.
+            %% Map association order is meaningful.
             %% See tests/typechecker_tests.erl:622 (deep_normalize)
             %% for a hacky solution for tests.
-            ret(type(map, Assocs1 ++ Assocs2))
+            MAssocs1 = maps:from_list([ {Key, As} || ?type(_, [Key, _]) = As <- Assocs1 ]),
+            MAssocs2 = maps:from_list([ {Key, As} || ?type(_, [Key, _]) = As <- Assocs2 ]),
+            %% TODO: repeated, extract to external fun
+            IsMandatory = fun
+                              (?type(map_field_exact, _)) -> true;
+                              (_) -> false
+                          end,
+            MandatoryKeys0 = [ Key || ?type(_, [Key, _]) <- lists:filter(IsMandatory, Assocs1 ++ Assocs2) ],
+            MandatoryKeys = lists:usort(MandatoryKeys0),
+            {Assocs, Css} = lists:unzip(lists:map(fun (Key) -> glb_ty_map_step(Key, MAssocs1, MAssocs2, A, TEnv) end,
+                                                  MandatoryKeys)),
```

Comment:
Please add some comments to explain what's going on here. I'm sure you know exactly what you're doing right now, but look at it again next year and you will be as lost as I am right now. :-)